src/gennylib/include/gennylib/parallel.hpp

```diff
@@ -0,0 +1,245 @@
+ *
+ * This container object also meets the BasicLockable requirement.
+ */
+template <typename T>
+class AtomicContainer {
+public:   
+    using value_type = typename T::value_type;
+
+
+    AtomicContainer() = default;
+    AtomicContainer(const AtomicContainer& rhs) {
+        std::lock_guard<lock_t> lock(rhs._mutex);
+        this->_container = rhs._container;
+    }
+    AtomicContainer(AtomicContainer&& rhs) {
+        std::lock_guard<lock_t> lock(rhs._mutex);
+        this->_container = std::move(rhs._container);
+    }
+
+    AtomicContainer<T>& operator=(const AtomicContainer& rhs) {
+        std::lock_guard<lock_t> lock(rhs._mutex);
+        this->_container = rhs._container;
+        return *this;
+    }
+    AtomicContainer<T>& operator=(AtomicContainer&& rhs) {
+        std::lock_guard<lock_t> lock(rhs._mutex);
+        this->_container = std::move(rhs._container);
+    }
+
+    // STL Methods
+    auto&& front() {
+        std::lock_guard<lock_t> lock(_mutex);
+        return _container.front();
+    }
+    template <typename... Args>
+    auto&& emplace_back(Args... args) {
+        std::lock_guard<lock_t> lock(_mutex);
+        return _container.emplace_back(std::forward<Args>(args)...);
+    }
+    void push_back(const typename T::value_type& value) {
+        std::lock_guard<lock_t> lock(_mutex);
+        _container.push_back(value);
+    }
+    void push_back(typename T::value_type&& value) {
+        std::lock_guard<lock_t> lock(_mutex);
+        _container.push_back(std::move(value));
+    }
+    auto&& back() {
+        std::lock_guard<lock_t> lock(_mutex);
+        return _container.back();
+    }
+    auto&& at(size_t pos) {
+        std::lock_guard<lock_t> lock(_mutex);
+        return _container.at(pos);
+    }
+    void pop_front() {
+        std::lock_guard<lock_t> lock(_mutex);
+        _container.pop_front();
+        return;
+    }
+    size_t size() const {
+        std::lock_guard<lock_t> lock(_mutex);
+        return _container.size();
+    }
+    bool empty() const {
+        std::lock_guard<lock_t> lock(_mutex);
+        return _container.empty();
+    }
+
+    const auto& operator[](size_t pos) const {
+        std::lock_guard<lock_t> lock(_mutex);
+        return _container[pos];
+    }
+    auto& operator[](size_t pos) {
+        std::lock_guard<lock_t> lock(_mutex);
+        return _container[pos];
+    }
+
+    /**
+     * It's possibly abusive to be treating these as const, but we want to be able to do
+     * locked iteration in const contexts, and we treat the locks as mutable elsewhere.
+     */
+    void lock() const {
+        // Paranoidly hold the lock here in case there are shenanigans
+        // between the flag flip and the unlock.
+        std::lock_guard<lock_t> lock(_mutex);
+        if (_forciblyHeld) {
+            BOOST_THROW_EXCEPTION(
+                std::logic_error("Cannot forcibly hold AtomicContainer lock multiple times."));
+        }
+        _mutex.lock();
```

Comment:
Since we're using `std::recursive_mutex`, my understanding is that at the end of this function, the `_mutex` has a lock count of two. When the guard is destructed the `unlock()` is called so when the function returns the lock count is one (representing the lock held by the caller).

We're effectively using the recursive mutex to allow multiple lockings to simulate a non-recursive lockable object with atomic operations.